# Credit Card Customer Churn - ETL

## Objectives

 - Load the raw credit card customer churn dataset.
- Understand the structure and quality of the data.
- Identify missing values, duplicate records, unsuitable data types, and non-informative features.
- Apply only the transformations required to prepare the dataset for analysis.
- Save a clean processed dataset for EDA, statistical analysis, machine learning, and Power BI.


## Inputs

- `data/raw/BankChurners.csv`
## Outputs

- `data/processed/bank_churners_clean.csv`



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\CodeInstitute\\credit-card-customer-churn-analysis\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("Working directory changed to project root.")

Working directory changed to project root.


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\CodeInstitute\\credit-card-customer-churn-analysis'

# Section 1. Load the Raw Dataset

The raw dataset is loaded from the `data/raw` directory.  
At this stage, no transformations are applied so the original dataset remains unchanged.

In [4]:
import pandas as pd
import numpy as np

df_raw = pd.read_csv("data/raw/BankChurners.csv")

df_raw.head()

,CLIENTNUM,Attrition_Flag,Customer_Age,Gender,Dependent_count,Education_Level,Marital_Status,Income_Category,Card_Category,Months_on_book,...,Credit_Limit,Total_Revolving_Bal,Avg_Open_To_Buy,Total_Amt_Chng_Q4_Q1,Total_Trans_Amt,Total_Trans_Ct,Total_Ct_Chng_Q4_Q1,Avg_Utilization_Ratio,Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1,Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2
0,768805383,Existing Customer,45,M,3,High School,Married,$60K - $80K,Blue,39,...,12691.0,777,11914.0,1.335,1144,42,1.625,0.061,0.000093,0.99991
1,818770008,Existing Customer,49,F,5,Graduate,Single,Less than $40K,Blue,44,...,8256.0,864,7392.0,1.541,1291,33,3.714,0.105,0.000057,0.99994
2,713982108,Existing Customer,51,M,3,Graduate,Married,$80K - $120K,Blue,36,...,3418.0,0,3418.0,2.594,1887,20,2.333,0.000,0.000021,0.99998
3,769911858,Existing Customer,40,F,4,High School,Unknown,Less than $40K,Blue,34,...,3313.0,2517,796.0,1.405,1171,20,2.333,0.760,0.000134,0.99987
4,709106358,Existing Customer,40,M,3,Uneducated,Married,$60K - $80K,Blue,21,...,4716.0,0,4716.0,2.175,816,28,2.500,0.000,0.000022,0.99998


---

# Section 2. Data Understanding

This section explores the structure and basic characteristics of the dataset before any cleaning or transformation is applied.

The following checks are performed:
- Dataset dimensions
- Column names
- Data types
- Sample records
- Summary statistics
- Target variable distribution

In [5]:
print(f"Dataset shape: {df_raw.shape}")

Dataset shape: (10127, 23)


In [6]:
df_raw.columns.tolist()

['CLIENTNUM',
 'Attrition_Flag',
 'Customer_Age',
 'Gender',
 'Dependent_count',
 'Education_Level',
 'Marital_Status',
 'Income_Category',
 'Card_Category',
 'Months_on_book',
 'Total_Relationship_Count',
 'Months_Inactive_12_mon',
 'Contacts_Count_12_mon',
 'Credit_Limit',
 'Total_Revolving_Bal',
 'Avg_Open_To_Buy',
 'Total_Amt_Chng_Q4_Q1',
 'Total_Trans_Amt',
 'Total_Trans_Ct',
 'Total_Ct_Chng_Q4_Q1',
 'Avg_Utilization_Ratio',
 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1',
 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2']

In [7]:
df_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10127 non-null  int64  
 1   Attrition_Flag                                                                                                                      10127 non-null  str    
 2   Customer_Age                                                                                                                        10127 non-null  int64  
 3   Gender                                                                                      

In [8]:
df_raw.describe().T

,count,mean,std,min,25%,50%,75%,max
CLIENTNUM,10127.0,7.391776e+08,3.690378e+07,7.080821e+08,7.130368e+08,7.179264e+08,7.731435e+08,8.283431e+08
Customer_Age,10127.0,4.632596e+01,8.016814e+00,2.600000e+01,4.100000e+01,4.600000e+01,5.200000e+01,7.300000e+01
Dependent_count,10127.0,2.346203e+00,1.298908e+00,0.000000e+00,1.000000e+00,2.000000e+00,3.000000e+00,5.000000e+00
Months_on_book,10127.0,3.592841e+01,7.986416e+00,1.300000e+01,3.100000e+01,3.600000e+01,4.000000e+01,5.600000e+01
Total_Relationship_Count,10127.0,3.812580e+00,1.554408e+00,1.000000e+00,3.000000e+00,4.000000e+00,5.000000e+00,6.000000e+00
Months_Inactive_12_mon,10127.0,2.341167e+00,1.010622e+00,0.000000e+00,2.000000e+00,2.000000e+00,3.000000e+00,6.000000e+00
Contacts_Count_12_mon,10127.0,2.455317e+00,1.106225e+00,0.000000e+00,2.000000e+00,2.000000e+00,3.000000e+00,6.000000e+00
Credit_Limit,10127.0,8.631954e+03,9.088777e+03,1.438300e+03,2.555000e+03,4.549000e+03,1.106750e+04,3.451600e+04
Total_Revolving_Bal,10127.0,1.162814e+03,8.149873e+02,0.000000e+00,3.590000e+02,1.276000e+03,1.784000e+03,2.517000e+03
Avg_Open_To_Buy,10127.0,7.469140e+03,9.090685e+03,3.000000e+00,1.324500e+03,3.474000e+03,9.859000e+03,3.451600e+04


In [9]:
df_raw["Attrition_Flag"].value_counts()

Attrition_Flag
Existing Customer    8500
Attrited Customer    1627
Name: count, dtype: int64

In [10]:
df_raw["Attrition_Flag"].value_counts(normalize=True) * 100

Attrition_Flag
Existing Customer    83.934038
Attrited Customer    16.065962
Name: proportion, dtype: float64

---

# Section 3. Data Quality Assessment

In this section, we assess the quality of the raw dataset before applying any cleaning transformations.

We will check:
- Missing values
- Duplicate records
- Unique values
- Potential data quality issues

In [11]:
# Check missing values
df_raw.isnull().sum()

CLIENTNUM                                                                                                                             0
Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                 

In [12]:
# Check duplicate records
df_raw.duplicated().sum()

np.int64(0)

In [13]:
# Check unique values in categorical columns
categorical_cols = df_raw.select_dtypes(include="object").columns

for col in categorical_cols:
    print(f"\n{col}:")
    print(df_raw[col].value_counts())

C:\Users\lap-mohammadnejhad\AppData\Local\Temp\ipykernel_9416\2489443431.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df_raw.select_dtypes(include="object").columns



Attrition_Flag:
Attrition_Flag
Existing Customer    8500
Attrited Customer    1627
Name: count, dtype: int64

Gender:
Gender
F    5358
M    4769
Name: count, dtype: int64

Education_Level:
Education_Level
Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: count, dtype: int64

Marital_Status:
Marital_Status
Married     4687
Single      3943
Unknown      749
Divorced     748
Name: count, dtype: int64

Income_Category:
Income_Category
Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: count, dtype: int64

Card_Category:
Card_Category
Blue        9436
Silver       555
Gold         116
Platinum      20
Name: count, dtype: int64


In [14]:
# Count "Unknown" values in categorical columns
for col in categorical_cols:
    unknown_count = (df_raw[col] == "Unknown").sum()
    if unknown_count > 0:
        print(col, ":", unknown_count)


Education_Level : 1519
Marital_Status : 749
Income_Category : 1112


In [15]:

# Identify unnecessary Naive Bayes columns
naive_bayes_cols = [
    col for col in df_raw.columns
    if col.startswith("Naive_Bayes_Classifier")
]

naive_bayes_cols


['Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1',
 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2']

---

# Section 4. Data Cleaning

In this section, the dataset is cleaned by removing unnecessary columns and preparing categorical values for further analysis.

In [16]:
# Create a copy of the raw dataset for cleaning
df_clean = df_raw.copy()

df_clean.shape


(10127, 23)

In [17]:
# Remove unnecessary Naive Bayes classifier columns
df_clean = df_clean.drop(columns=naive_bayes_cols)

# Check the new dataset shape
df_clean.shape


(10127, 21)

### Handling "Unknown" Categories

The categorical variables Education_Level, Marital_Status, and Income_Category contain "Unknown" values.

These records are retained rather than removed to avoid unnecessary data loss. The "Unknown" category may also provide useful information during exploratory data analysis.

In [18]:
# Confirm "Unknown" values are retained
for col in ["Education_Level", "Marital_Status", "Income_Category"]:
    print(col, ":", (df_clean[col] == "Unknown").sum())

Education_Level : 1519
Marital_Status : 749
Income_Category : 1112


### Customer Identifier

CLIENTNUM is a unique customer identifier. It is retained in the cleaned dataset for record identification but will not be used as a predictive feature in the machine learning model.


In [19]:
# Check whether CLIENTNUM uniquely identifies each customer
print("Total rows:", len(df_clean))
print("Unique CLIENTNUM:", df_clean["CLIENTNUM"].nunique())


Total rows: 10127
Unique CLIENTNUM: 10127


### Create Churn Target Variable

A binary Churn variable is created from Attrition_Flag to simplify analysis and prepare the dataset for machine learning.

- 0 = Existing Customer
- 1 = Attrited Customer


In [20]:
# Create binary churn target
df_clean["Churn"] = df_clean["Attrition_Flag"].map({
    "Existing Customer": 0,
    "Attrited Customer": 1
})

# Check the result
df_clean[["Attrition_Flag", "Churn"]].value_counts()


Attrition_Flag     Churn
Existing Customer  0        8500
Attrited Customer  1        1627
Name: count, dtype: int64

---

# Section 5. Prepare Dataset for Analysis

The cleaned dataset is prepared for exploratory data analysis and subsequent machine learning.


In [21]:
# Remove the original target column
df_processed = df_clean.drop(columns=["Attrition_Flag"])

df_processed.shape

(10127, 21)

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Section 6. Save Processed Dataset


The cleaned dataset is saved in the processed data directory for use in exploratory data analysis and machine learning.


In [22]:
# Save the processed dataset
df_processed.to_csv(
    "data/processed/bank_churners_clean.csv",
    index=False
)


In [23]:
import os

os.path.exists("data/processed/bank_churners_clean.csv")


True

---

# Conclusion

The ETL process prepared the Credit Card Customer Churn dataset for further analysis.

### Key Outcomes

- The raw dataset contained 10,127 customer records and 23 columns.
- No missing values or duplicate records were identified.
- "Unknown" categories were identified in `Education_Level`, `Marital_Status`, and `Income_Category` and were retained to avoid unnecessary data loss.
- Two pre-generated Naive Bayes classifier columns were removed because they are not required for this analysis.
- `CLIENTNUM` was retained as a customer identifier but will not be used as a predictive feature.
- A binary `Churn` target was created, where `0` represents an existing customer and `1` represents an attrited customer.
- The original `Attrition_Flag` column was removed after creating the binary target.
- The processed dataset was saved as `data/processed/bank_churners_clean.csv`.
